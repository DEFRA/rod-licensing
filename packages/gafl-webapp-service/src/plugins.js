import Inert from '@hapi/inert'
import Vision from '@hapi/vision'
import Disinfect from 'disinfect'
import Scooter from '@hapi/scooter'
import Blankie from 'blankie'
import Crumb from '@hapi/crumb'
import HapiGapi from '@defra/hapi-gapi'
import Cookie from '@hapi/cookie'
import HapiI18n from 'hapi-i18n'
import { useSessionCookie } from './session-cache/session-manager.js'
import { UTM } from './constants.js'
import { getCsrfTokenCookieName } from './server.js'
import Dirname from '../dirname.cjs'
import path from 'path'

// This is a hash of the inline script at line 31 of the GDS template. It is added to the CSP to except the in-line
// script. It needs the quotes.
const scriptHash = "'sha256-+6WnXIl4mbFTCARd8N3COQmT3bJJmo32N8q8ZSQAIcU='"

const initialiseDisinfectPlugin = () => ({
  plugin: Disinfect,
  options: {
    disinfectQuery: true,
    disinfectParams: true,
    disinfectPayload: true
  }
})

const initialiseBlankiePlugin = () => ({
  plugin: Blankie,
  options: {
    /*
     * This defines the content security policy - which is as restrictive as possible
     * It must allow web-fonts from 'fonts.gstatic.com'
     */
    fontSrc: ['self', 'fonts.gstatic.com', 'data:'],
    scriptSrc: ['self', 'unsafe-inline', scriptHash],
    generateNonces: true,
    frameAncestors: 'none'
  }
})

const initialiseCrumbPlugin = () => ({
  plugin: Crumb,
  options: {
    key: getCsrfTokenCookieName(),
    cookieOptions: {
      isSecure: process.env.NODE_ENV !== 'development',
      isHttpOnly: process.env.NODE_ENV !== 'development'
    },
    logUnauthorized: true
  }
})

const initialiseHapiGapiPlugin = () => {
  const hapiGapiPropertySettings = []
  if (process.env.ANALYTICS_PRIMARY_PROPERTY) {
    hapiGapiPropertySettings.push({ id: process.env.ANALYTICS_PRIMARY_PROPERTY, hitTypes: ['pageview', 'event', 'ecommerce'] })
  } else {
    console.warn("ANALYTICS_PRIMARY_PROPERTY not set, so Google Analytics won't track this")
  }
  if (process.env.ANALYTICS_XGOV_PROPERTY) {
    hapiGapiPropertySettings.push({ id: process.env.ANALYTICS_XGOV_PROPERTY, hitTypes: ['pageview'] })
  } else {
    console.warn("ANALYTICS_XGOV_PROPERTY not set, so Google Analytics won't track this")
  }

  return {
    plugin: HapiGapi,
    options: {
      propertySettings: hapiGapiPropertySettings,
      sessionIdProducer: async request => {
        let sessionId = null
        if (useSessionCookie(request)) {
          const { gaClientId } = await request.cache().helpers.status.get()
          sessionId = gaClientId ?? (await request.cache().getId())
        }
        return sessionId
      },
      attributionProducer: async request => {
        if (useSessionCookie(request)) {
          const { attribution } = await request.cache().helpers.status.get()

          if (attribution) {
            return {
              campaign: attribution[UTM.CAMPAIGN],
              content: attribution[UTM.CONTENT],
              medium: attribution[UTM.MEDIUM],
              source: attribution[UTM.SOURCE],
              term: attribution[UTM.TERM]
            }
          }
        }
        return {}
      }
    }
  }
}

const initialiseHapiI18nPlugin = () => {
  const showWelshContent = process.env.SHOW_WELSH_CONTENT?.toLowerCase() === 'true'
  return {
    plugin: HapiI18n,
    options: {
      locales: ['en', ...(showWelshContent ? ['cy'] : [])],
      directory: path.join(Dirname, 'src/locales'),
      ...(showWelshContent && { queryParameter: 'lang' })
    }
  }
}

export const getPlugins = () => {
  return [
    Inert,
    Vision,
    Scooter,
    Cookie,
    initialiseDisinfectPlugin(),
    initialiseBlankiePlugin(),
    initialiseCrumbPlugin(),
    initialiseHapiGapiPlugin(),
    initialiseHapiI18nPlugin()
  ]
}
