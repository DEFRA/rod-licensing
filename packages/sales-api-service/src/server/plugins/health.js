import HapiAndHealthy from 'hapi-and-healthy'
import { dynamicsClient } from '@defra-fish/dynamics-lib'
import Project from '../../project.cjs'
import { AWS } from '@defra-fish/connectors-lib'
const { ddb, sqs } = AWS()

export default {
  plugin: HapiAndHealthy,
  options: {
    id: Project.packageJson.name,
    env: process.env.NODE_ENV,
    name: Project.packageJson.description,
    version: Project.packageJson.version,
    custom: {
      versions: process.versions,
      dependencies: Project.packageJson.dependencies
    },
    tags: ['api', 'health'],
    test: {
      node: [
        async () => {
          return { connection: 'dynamics', status: 'ok', ...(await dynamicsClient.executeUnboundFunction('RetrieveVersion')) }
        },
        async () => {
          return { connection: 'dynamodb', status: 'ok', ...(await ddb.listTables()) }
        },
        async () => {
          return { connection: 'sqs', status: 'ok', ...(await sqs.listQueues()) }
        }
      ]
    }
  }
}
