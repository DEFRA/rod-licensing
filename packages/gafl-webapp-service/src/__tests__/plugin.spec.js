import { getPlugins } from '../plugins'

jest.mock('../constants', () => ({
  CommonResults: {
    OK: 'ok'
  },
  ShowDigitalLicencePages: {
    YES: 'show-digital-licence-yes'
  }
}))

jest.mock('../server.js', () => ({
  getCsrfTokenCookieName: jest.fn()
}))
jest.mock('debug', () => jest.fn(() => jest.fn()))

describe('plugins', () => {
  describe('initialiseHapiI18nPlugin', () => {
    it.each([
      [['en'], undefined, undefined],
      [['en'], undefined, false],
      [['en'], undefined, 'false'],
      [['en', 'cy'], 'lang', true],
      [['en', 'cy'], 'lang', 'true']
    ])(
      'should return the locale as %s and queryParameter as %s, if SHOW_WELSH_CONTENT is %s',
      async (expectedLocale, expectedQueryParamter, showWelshContent) => {
        process.env.SHOW_WELSH_CONTENT = showWelshContent
        const hapI18nPlugin = getPlugins().find(plugin => plugin?.plugin?.plugin?.name === 'hapi-i18n')
        expect(hapI18nPlugin.options.locales).toStrictEqual(expectedLocale)
        expect(hapI18nPlugin.options.queryParameter).toBe(expectedQueryParamter)
      }
    )
  })

  describe('initialiseBlankiePlugin', () => {
    it('should return expecetd src details', async () => {
      const blankiePlugin = getPlugins().find(plugin => plugin?.plugin?.plugin?.pkg?.name === 'blankie')
      expect(blankiePlugin.options).toMatchSnapshot()
    })
  })
})
