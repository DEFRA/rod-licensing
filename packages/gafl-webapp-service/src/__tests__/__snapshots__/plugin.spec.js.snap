// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins initialiseBlankiePlugin should return expecetd src details 1`] = `
Object {
  "connectSrc": Array [
    "self",
    "*.google-analytics.com",
    "*.googletagmanager.com",
    "*.analytics.google.com",
  ],
  "fontSrc": Array [
    "self",
    "fonts.gstatic.com",
    "data:",
  ],
  "frameAncestors": "none",
  "generateNonces": true,
  "imgSrc": Array [
    "self",
    "unsafe-inline",
    "*.googletagmanager.com",
    "fonts.gstatic.com",
    "data:",
    "https://ssl.gstatic.com/",
  ],
  "manifestSrc": Array [
    "self",
  ],
  "scriptSrc": Array [
    "self",
    "unsafe-inline",
    "'sha256-GUQ5ad8JK5KmEWmROf3LZd9ge94daqNvd8xy9YS1iDw='",
    "*.googletagmanager.com",
    "*.tagassistant.google.com",
    "unsafe-eval",
    "https://tagmanager.google.com/",
  ],
  "styleSrc": Array [
    "unsafe-inline",
    "self",
    "fonts.googleapis.com",
    "*.googletagmanager.com",
    "https://tagmanager.google.com/",
  ],
}
`;
